Ray casting returns uv coordinates

rfm1201 opened the pull request:
Mesh.js raycast method is updated to calculate intersection point in uv
coordinates. The calculus is done in the new inner method
textureIntersection and added to the object returned as the uv property.

Texture.js has a new public method, transformUv, that convert the
coordinates from uv (range 0 to 1) to the real texture coordinates. The
input parameter is modified on output. Useful to have the intersection
point in texture coordinates.

There is no example in this commit because I only have example with
canvas, no with plain image.

Fixes #6762

rfm1201 committed:
Ray casting gives texture intersection in uv coordinates

Mesh.js raycast method is updated to calculate intersection point in uv
coordinates. The calculus is done in the new inner method
textureIntersection and added to the object returned as the uv property.

Texture.js has a new public method, transformUv, that convert the
coordinates from uv (range 0 to 1) to the real texture coordinates. The
input parameter is modified on output. Useful to have the intersection
point in texture coordinates.

There is no example in this commit because I only have example with
canvas, no with plain image.

WestLangley commented on src/objects/Mesh.js:
```diff
@@ -107,6 +111,21 @@ THREE.Mesh.prototype.raycast = ( function () {
+		var textureIntersection = function ( pIntersection, p1, p2, p3, uv1, uv2, uv3 ) {
+
+			var bary = THREE.Triangle.barycoordFromPoint( pIntersection, p1, p2, p3 );
+
```
`THREE.Triangle.barycoordFromPoint()` instantiates a new `Vector3()` with each call. To prevent that, you need to pass in an `optionalTarget`.

You can use the prototype method with the same name, instead, if you want.

WestLangley commented on src/objects/Mesh.js:
```diff
@@ -152,6 +171,16 @@ THREE.Mesh.prototype.raycast = ( function () {
+							uvC.fromArray( uvs, c * 2 );
+							uv = textureIntersection( intersectionPoint, vA, vB, vC, uvA, uvB, uvC )
+						}
+
```
Defer the calculation of UVs until after the near plane / far plane test.

WestLangley commented on src/objects/Mesh.js:
```diff
@@ -152,6 +171,16 @@ THREE.Mesh.prototype.raycast = ( function () {
+							uvA.fromArray( uvs, a * 2 );
+							uvB.fromArray( uvs, b * 2 );
+							uvC.fromArray( uvs, c * 2 );
+							uv = textureIntersection( intersectionPoint, vA, vB, vC, uvA, uvB, uvC )
```
semicolon missing (3 instances in this file)

WestLangley commented:
Nice job!

> There is no example in this commit because I only have example with
canvas, no with plain image.

Fine, but did you test all logic paths in your new code to confirm it is working properly?

rfm1201 committed:
With WestLangley suggestions

Add a new variable to avoid instantiation of a Vector3 in each call to
THREE.Triangle.barycoordFromPoint().

Defer calculus after the near plane / far plane test. Need to do a copy
of intersectionPoint due to the applyMatrix4.

Test with a PlaneBufferGeometry and a BoxGeometry. Didn’t test with
indexed BufferGeometry (don’t have any in my program) but I’m quite
comfortable the uvA, uvB and uvC are the right one in that case.

rfm1201 commented:
I add the modifications suggested and commit.

I don't know if I have something else to do to update the pull request.

rfm1201 commented:
I'm working on a demo page with the different kind of geometries. I'm on my way, but I realize that I made a mistake in the description of the transformUv method I added to the Texture class.

It doesn’t convert in real texture coordinates. It was my initial goal but I changed my mind (and forget it since June).

So it converts the uv given by the ray cast to uv in the [0, 1] space, according to the mapping of the texture. The caller has to convert in the texture coordinates by multiplying u by the image (or canvas) width and v by the height.

kylegmaxwell commented:
You generally need to git push so the remote server gets the new commit.

On Fri, Aug 21, 2015 at 9:34 AM rfm1201 <notifications@github.com> wrote:

> I'm working on a demo page with the different kind of geometries. I'm on
> my way, but I realize that I made a mistake in the description of the
> transformUv method I added to the Texture class.
>
> It doesn’t convert in real texture coordinates. It was my initial goal but
> I changed my mind (and forget it since June).
>
> So it converts the uv given by the ray cast to uv in the [0, 1] space,
> according to the mapping of the texture. The caller has to convert in the
> texture coordinates by multiplying u by the image (or canvas) width and v
> by the height.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mrdoob/three.js/pull/7023#issuecomment-133483844>.
>

rfm1201 committed:
A page to test or demonstrate

The page tests the 3 intersection calculus possible.
The cube as a normal geometry, the plane a buffered geometry and the
circle a buffered geometry with index.

rfm1201 commented:
The third commit add a web page, webgl_texture_raycast.html, to test and/or demonstrate the ray cast on texture.

All meshes use their own texture, with different mapping, but all textures are bound to the same canvas.
When you click on a face, a cross is drawn in the canvas where you clicked.

WestLangley commented:
This looks good.

1. Can you also demonstrate that is works with `texture.repeat.set( a, b )` and `texture.offset.set( c, d )`?
2. Also with `ClampToEdgeMapping`.
3. I find that if you remove the black cross and change the white one to yellow, the cross is much easier to see. (Just a suggestion)
4. [mrdoobapproves](http://zz85.github.io/mrdoobapproves/) can help you with your formatting.

I like what you have done here.

rfm1201 commented:
Ok, I will add the different texture configuration and refactor the code, but only tomorrow.

rfm1201 committed:
Add tuning to the web page

The example web page now allows the change of different properties of
the circle's texture:
- wrapS
- wrapT
- offset
- repeat

The change on the js files are only formatting ones (mrdoobapproves)

rfm1201 committed:
refactor web page code

The web page code is organized as most of the other example web pages.

rfm1201 commented:
Oups, I forget to tell that I have done what I think should be done.

So with this version of the web test page, the user may change wrapS, wrapT, offset and repeat on the right mesh (the circle, indexed buffer).
No change is allowed on the other meshes, because they are too small, and the values that can be changed are not concerned by the way the geometry is stored. If it works on one, it works on all.

WestLangley commented on examples/webgl_texture_raycast.html:
```diff
@@ -0,0 +1,377 @@
+
+					that._rayonviseur = Math.ceil(Math.min(that._canvas.width, that._canvas.height / 30));
+					that._exterieurCroix = Math.ceil(0.70710678 * that._rayonviseur);
+					that._interieurCroix = Math.ceil(that._exterieurCroix / 10);
```
English, please.

WestLangley commented on examples/webgl_texture_raycast.html:
```diff
@@ -0,0 +1,377 @@
+				var uvs = circleGeometry.attributes.uv.array;
+				for (var i = 0; i < uvs.length; i++) {
+
+					uvs[i] = (uvs[i] - 0.25) * 2;
```
I would suggest you remove this. `ClampToEdgeWrapping` does not look right.

rfm1201 replied:
I will, but what do you means by "it does not look right"?
It is rendered correctly, and I think that the way I calculate the intersection is the right one. For me it works as I expected.
If you mean, the picture isn’t very pretty, I think you are right, but I always think ClampToEdgeWrapping does ugly picture. I just (try to) prove it works.

WestLangley replied:
Well, `ClampToEdgeWrapping` does not really make sense if UVs exceed 1. I guess it depends on whether @mrdoob plans to include the demo in the examples.

WestLangley commented:
1. You have power-of-two warnings in the console.

2. [mrdoobapproves](http://zz85.github.io/mrdoobapproves/) can help you with your formatting.

rfm1201 committed:
Full English, no more console warning

No more French.
Power-of-two warnings checked.
Default for the circle is now RepeatWrapping.

rfm1201 commented:
Full English (I hope), no more console warning and RepeatWrapping as default for the circle.

rfm1201 commented:
I get the current three.js dev branch this morning and Mesh.js had been updated.
What am I supposed to do, leave my Mesh.js code as it currently is or put the modified lines?

gero3 commented:
![image](https://cloud.githubusercontent.com/assets/155535/9546155/a8218230-4d90-11e5-9f54-42fe0536589b.png)

As long as above is green, you don't have to worry about your code.

mrdoob commented:
If I'm not mistaken we already have the barycentric coordinates (`DdQxE2` and `DdE1xQ`) when we're doing `ray.intersectTriangle()`, no?

https://github.com/mrdoob/three.js/blob/dev/src/math/Ray.js#L475-L496

So, instead of calculating them again we could add a `bari` parameter to `intersectTriangle()` so we could store them.

Also, I noticed that we're not using `THREE.Triangle.barycoordFromPoint()` anywhere at the moment.

WestLangley commented:
> Also, I noticed that we're not using THREE.Triangle.barycoordFromPoint() anywhere at the moment.

`THREE.Triangle.prototype.barycoordFromPoint()` calls the static method.

> So, instead of calculating them again we could add a bari parameter to intersectTriangle() so we could store them.

The UV calculation is a rare event, so maybe the code is fine for now?

mrdoob commented:
>> Also, I noticed that we're not using THREE.Triangle.barycoordFromPoint() anywhere at the moment.

> `THREE.Triangle.prototype.barycoordFromPoint()` calls the static method.

We're not calling `THREE.Triangle.prototype.barycoordFromPoint()` either 😁 (I mean, until now...)

>> So, instead of calculating them again we could add a bari parameter to intersectTriangle() so we could store them.

> The UV calculation is a rare event, so maybe the code is fine for now?

Yeah, I guess I can start by merging this and try to optimise later.

mrdoob commented:
Many thanks!

eudorias commented:
The UV calculation is a rare event, so maybe the code is fine for now?

-Nooo , i need soo badly to find texture pixel coordinates from intersects uv ...

threejsworker commented:
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#7023
